README.md

```diff
@@ -1 +1,292 @@
+
+However, there's a limitation to this approach. Let's say we need to have three sets of API keys: one for development, one for testing and one for production. You can write `if/else` conditions in `main.js` but then you end up repeating the same logic over and over in other files which need to use the configurations.
+This approach isn't great because we end up with a lot of logic which really belongs to the configuration module spread out across various main files (files which import the module). Take a look at the code below (also in `main-envirionment-bad.js`):
+
+```js
+if (process.env.NODE_ENV == 'production') {
+  var configs = require('./configs-production')
+} else if (process.env.NODE_ENV == 'testing') {
+  var configs = require('./configs-testing')
+} else { // Assume development
+  var configs = require('./configs')
+}
+
+console.log('URL is ' + configs.url)
+```
+
+If you are thinking we should move the logic to the `config.js` file, the you are right. But how do we go about it if it's just an object?
+
+## Exporting Functions
+
+To make the module smarter (more dynamic), we need to use the `module.exports = function() {}` pattern which is more versatile. It allows us to execute some logic and even use parameters from the function's arguments.
+
+With the function pattern, we can re-write our configuration module to do more things. In other words, the module will can behave differently because we can put some logic in its code. This is `environment.js` file. The first thing in it is getting the environment or using the one from the environment variable `NODE_ENV` if the argument is not set:
+
+```js
+// environment.js
+module.exports = function(env) {
+  env = env || process.env.NODE_ENV
+```
+
+Then we define two conditions and return different configurations accordingly (in this example only API keys are different):
+
+```js
+  if (env == 'production') {
+    return {
+      url: 'http://webapplog.com',
+      name: 'React Quicly',
+      port: 3000,
+      apiKey: "34EC5CE9-CD02-4C4E-B5D9-D7CEF2F12F2C"
+    }
+  } else if (env == 'testing') {
+    return {
+      url: 'http://webapplog.com',
+      name: 'React Quicly',
+      port: 3000,
+      apiKey: "D2230FC4-7E2A-4003-9372-0BF80107558D"
+    }
+  } else { // Assume development
+    return {
+      url: 'http://webapplog.com',
+      name: 'React Quicly',
+      port: 3000,
+      apiKey: "D1FCFE58-575F-40FF-AF36-0CBD6A367756"
+    }
+  }
+}
+```
+
+The `main-environment.js` file can pass the argument or not, but it needs to invoke the imported configs method (from `environment`):
+
+```js
+// main-environment.js
+var configsInit = require('./environment')
+configs = configsInit('your-awesome-environment')
+console.log('URL is ' + configs.url)
+console.log('API key is ' +  configs.apiKey)
+```
+
+or you can merge 1st and 2nd lines into a single statement (more common style):
+
+```js
+var configs = require('./environment')()
```

Comment:
This could potentially be confusing to the user. We should emphasize that `require('./environment')` makes their function available immediately for invocation, so we can call it right away.